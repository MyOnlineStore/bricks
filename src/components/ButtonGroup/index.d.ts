import { FunctionComponent } from 'react';
declare const ButtonGroup: FunctionComponent;
export default ButtonGroup;
