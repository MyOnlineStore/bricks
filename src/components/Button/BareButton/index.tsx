import React, { Children, FunctionComponent } from 'react';
import StyledButton, { StyledAnchor } from './style';
import Icon, { MediumPropsType } from '../../Icon';
import Box from '../../Box';
import trbl from '../../../utility/trbl';
import Spinner from '../../Spinner';

type PropsType = {
    className?: string;
    compact?: boolean;
    title: string;
    variant?: 'primary' | 'destructive' | 'warning' | 'secondary' | 'plain';
    target?: HTMLAnchorElement['target'];
    href?: string;
    disabled?: boolean;
    icon?: MediumPropsType['icon'];
    iconAlign?: 'right' | 'left';
    id?: string;
    loading?: boolean;
    onClick?(): void;
};

const ButtonContents: FunctionComponent<PropsType> = (props): JSX.Element => (
    <>
        {props.loading && (
            <Box
                justifyContent="center"
                alignItems="center"
                position="absolute"
                left="0"
                top="0"
                right="0"
                bottom="0"
                padding={trbl(6)}
            >
                <Spinner color={'#fff'} />
            </Box>
        )}
        {props.icon && props.iconAlign !== 'right' && (
            <Box inline padding={trbl(0, 6, 0, 0)}>
                <Icon icon={props.icon} size={'medium'} />
            </Box>
        )}
        {Children.count(props.children) > 0 ? props.children : props.title}
        {props.icon && props.iconAlign === 'right' && (
            <Box inline padding={trbl(0, 0, 0, 6)}>
                <Icon icon={props.icon} size={'medium'} />
            </Box>
        )}
    </>
);

const BareButton: FunctionComponent<PropsType> = (props): JSX.Element => {
    const isLink = props.href !== undefined;

    const clickAction = (): void => {
        if (props.onClick !== undefined && props.disabled !== true && props.loading !== true) props.onClick();
    };

    if (isLink) {
        return (
            <StyledAnchor
                // @ts-ignore
                href={props.href}
                target={props.target}
                title={props.title}
                className={props.className}
                disabled={props.disabled}
                id={props.id}
            >
                <ButtonContents {...props}>{props.children}</ButtonContents>
            </StyledAnchor>
        );
    }

    return (
        <StyledButton
            title={props.title}
            className={props.className}
            onClick={clickAction}
            type="button"
            disabled={props.disabled}
            id={props.id}
        >
            <ButtonContents {...props}>{props.children}</ButtonContents>
        </StyledButton>
    );
};

export default BareButton;
export { PropsType };
