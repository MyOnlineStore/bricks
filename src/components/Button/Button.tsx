import React, { StatelessComponent } from 'react';
import { ThemeType } from '../../themes';
import styled from '../../utility/styled/styled';

type PropsType = {
    className?:string;
    title:string;
    type:'primary' | 'secondary' | 'destructive' | 'warning';
    target?:HTMLAnchorElement['target'];
    href?:string;
    action?():void;
};

const Button:StatelessComponent<PropsType> = (props):JSX.Element => {
    const Element = props.href !== undefined
        ? 'a'
        : 'button';

    const clickAction = ():void => {
        if (props.action !== undefined) {
            props.action();
        }
    };

    return (
        <Element
            type={props.type}
            title={props.title}
            className={props.className}
            href={props.href}
            target={props.target}
            onClick={clickAction}
        >
            {props.children}
        </Element>
    );
};

export default Button;
