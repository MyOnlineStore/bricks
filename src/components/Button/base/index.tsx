import React, { FC } from 'react';
import StyledButton, { StyledAnchor } from './style';

type PropsType = {
    className?: string;
    compact?: boolean;
    title: string;
    variant?: 'primary' | 'destructive' | 'warning' | 'secondary' | 'plain';
    target?: HTMLAnchorElement['target'];
    href?: string;
    disabled?: boolean;
    id?: string;
    loading?: boolean;
    'data-testid'?: string;
    onClick?(): void;
};

const ButtonBase: FC<PropsType> = (props): JSX.Element => {
    const isLink = props.href !== undefined;

    const clickAction = (): void => {
        if (props.onClick !== undefined && props.disabled !== true && props.loading !== true) props.onClick();
    };

    if (isLink) {
        return (
            <StyledAnchor
                // @ts-ignore
                href={props.href}
                target={props.target}
                title={props.title}
                className={props.className}
                disabled={props.disabled}
                id={props.id}
                loading={props.loading}
                data-testid={props['data-testid']}
            >
                {props.children}
            </StyledAnchor>
        );
    }

    return (
        <StyledButton
            title={props.title}
            className={props.className}
            onClick={clickAction}
            type="button"
            disabled={props.disabled}
            id={props.id}
            loading={props.loading}
            data-testid={props['data-testid']}
        >
            {props.children}
        </StyledButton>
    );
};

export default ButtonBase;
export { PropsType };
