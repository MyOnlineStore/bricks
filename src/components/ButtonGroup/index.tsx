import React, { Children, FunctionComponent } from 'react';
import Box from '../Box';

type PropsType = {
    stacked?: boolean;
    'data-testid'?: string;
};

const ButtonGroup: FunctionComponent<PropsType> = props => {
    const direction = props.stacked ? 'column' : 'row';
    const children = props.stacked ? Children.toArray(props.children) : Children.toArray(props.children).reverse();

    return (
        <Box
            direction={direction}
            justifyContent={props.stacked ? 'flex-start' : 'flex-end'}
            alignItems="stretch"
            wrap
            margin={[-6]}
            data-testid={props['data-testid']}
        >
            {children.map((child, index) => (
                <Box key={index} direction={direction} alignSelf="stretch" margin={[6]}>
                    {child}
                </Box>
            ))}
        </Box>
    );
};

export default ButtonGroup;
