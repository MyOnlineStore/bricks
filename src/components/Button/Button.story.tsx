import { boolean, select, text } from '@storybook/addon-knobs/react';
import { storiesOf } from '@storybook/react';
import React from 'react';
import styled from '../../utility/styled';
import trbl from '../../utility/trbl';
import Button from '../Button';
import ButtonGroup from '../ButtonGroup';
import Contrast from '../Contrast';
import Icon from '../Icon';
import Spacer from '../Spacer';

const StyledBackground = styled.div`
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background: #f8f9fb;
    padding: 24px;
`;

storiesOf('Button', module)
    .add('With text', () => {
        return (
            <Button
                variant={select(
                    'variant',
                    [
                        'primary',
                        'secondary',
                        'warning',
                        'destructive',
                        'flat'
                    ],
                    'primary'
                )}
                title={text('title', 'Click me')}
                disabled={boolean('disabled', false)}
                action={():void => { /* */ }}
                compact={boolean('compact', false)}
            />
        );
    })
    .add('With an icon', () => {
        return (
            <Button
                variant={select(
                    'variant',
                    [
                        'primary',
                        'secondary',
                        'warning',
                        'destructive',
                        'flat'
                    ],
                    'secondary'
                )}
                title={text('title', 'Click me')}
                disabled={boolean('disabled', false)}
                action={():void => { /* */ }}
                compact
            >
                <Icon size="small" icon="gear" />
            </Button>
        );
    })
    .add('On a contrast area', () => {
        return (
            <Contrast>
                <Spacer offsetType="inner" offset={trbl(12)}>
                    <ButtonGroup>
                        <Button
                            variant="secondary"
                            title={text('title', 'Click me')}
                            disabled={boolean('disabled', false)}
                            action={():void => { /* */ }}
                        />
                        <Button
                            variant="flat"
                            title={'Click me'}
                            action={():void => { /* */ }}
                            compact
                        >
                            <Icon size="small" icon="gear" />
                        </Button>
                    </ButtonGroup>
                </Spacer>
            </Contrast>
        );
    });
