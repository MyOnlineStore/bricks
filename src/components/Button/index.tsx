import React, { Children, ReactNode } from 'react';
import styled from '../../utility/_styled';
import Icon from '../Icon';
import BareButton, { PropsType as BaseButtonPropsType } from './base';
import Box from '../Box';
import Spinner from '../Spinner';

type CommonType = {
    backgroundColor: string;
    color: string;
    boxShadow: string;
};

type ComponentStateTypes = {
    idle: CommonType;
    hover: CommonType;
    focus: CommonType;
    active: CommonType;
};

type ButtonThemeType = {
    common: {
        borderRadius: string;
        borderWidth: string;
        fontWeight: string;
        fontFamily: string;
        fontSize: string;
        textDecoration: string;
    };
    primary: ComponentStateTypes;
    secondary: ComponentStateTypes;
    warning: ComponentStateTypes;
    destructive: ComponentStateTypes;
    plain: {
        hover: CommonType;
        focus: CommonType;
        active: CommonType;
        idle: {
            backgroundColor: string;
            color: string;
            boxShadow: string;
            textDecoration: string;
        };
    };
    disabled: {
        backgroundColor: string;
        color: string;
        stripingColor: string;
    };
};

type PropsType = BaseButtonPropsType & {
    loading?: boolean;
    variant: 'primary' | 'destructive' | 'warning' | 'secondary' | 'plain';
    compact?: boolean;
    disabled?: boolean;
    icon?: string;
    children?: ReactNode;
};

const Button = styled(BareButton).attrs((props: PropsType) => {
    const color = props.loading ? 'transparent' : undefined;

    const children: ReactNode = (
        <>
            {props.loading && (
                <Box
                    justifyContent="center"
                    alignItems="center"
                    position="absolute"
                    left="0"
                    top="0"
                    right="0"
                    bottom="0"
                    padding={[6]}
                >
                    <Spinner />
                </Box>
            )}
            {props.icon && (
                <Box inline padding={[0, 6, 0, 0]}>
                    <Icon size="medium" icon={props.icon} color={color} />
                </Box>
            )}
            <span style={{ color }}>{Children.count(props.children) > 0 ? props.children : props.title}</span>
        </>
    );

    return {
        children,
    };
})<PropsType>`
    ${({ theme, variant, compact, disabled, loading }): string => {
        const idle = `
            background-color: ${theme.Button[variant].idle.backgroundColor};
            box-shadow: ${theme.Button[variant].idle.boxShadow};
            color: ${theme.Button[variant].idle.color};
        `;

        const hover = `
            background-color: ${theme.Button[variant].hover.backgroundColor};
            box-shadow: ${theme.Button[variant].hover.boxShadow};
            color: ${theme.Button[variant].hover.color};
        `;

        const active = `
            transform: translateY(2px);
            background-color: ${theme.Button[variant].active.backgroundColor};
            box-shadow: ${theme.Button[variant].active.boxShadow};
            color: ${theme.Button[variant].active.color};
        `;

        const focus = `
            background-color: ${theme.Button[variant].focus.backgroundColor};
            box-shadow: ${theme.Button[variant].focus.boxShadow};
            color: ${theme.Button[variant].focus.color};
        `;

        return `
            ${idle}
            padding: 11px ${compact ? ' 12px' : '24px'};
            border-radius: ${theme.Button.common.borderRadius};
            ${variant === 'plain' ? `text-decoration: ${theme.Button.plain.idle.textDecoration}` : ''}

            &:hover {
                ${!loading && !disabled ? hover : idle}
            }

            &:focus {
                ${!loading && !disabled ? focus : idle}
            }

            &:active {
                ${!loading && !disabled ? active : idle}
            }

            &::before {
                opacity: ${disabled ? 1 : 0};
                ${disabled ? `color: ${theme.Button.disabled.color}` : 0};
                background: ${theme.Button.disabled.backgroundColor}
                    repeating-linear-gradient(
                        -45deg,
                        ${theme.Button.disabled.stripingColor},
                        ${theme.Button.disabled.stripingColor} 10px,
                        transparent 10px,
                        transparent 20px
                    );
                box-shadow: ${theme.Button[variant].idle.boxShadow};
                border-radius: ${theme.Button.common.borderRadius};
            }

            &:disabled {
                background: ${theme.Button.disabled.backgroundColor};
                color: ${theme.Button.disabled.color};
            }
            `;
    }};
`;

export default Button;
export { PropsType, ButtonThemeType };
