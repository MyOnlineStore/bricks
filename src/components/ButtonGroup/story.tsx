import { storiesOf } from '@storybook/react';
import { boolean } from '@storybook/addon-knobs';
import { withInfo } from '@storybook/addon-info';
import React from 'react';
import styled from 'styled-components';
import ButtonGroup from '.';
import Button from '../Button';

const Wrap = styled.div`
    max-width: 640px;
    border: solid 3px rgba(255, 36, 94, 0.3);
`;

storiesOf('ButtonGroup', module)
    .add('Default', () => {
        return (
            <Wrap>
                <ButtonGroup stacked={boolean('Stacked', false)}>
                    <Button variant="primary" title="Primary button">
                        Primary button
                    </Button>
                    <Button variant="secondary" title="Secondary button">
                        Secondary button
                    </Button>
                </ButtonGroup>
            </Wrap>
        );
    })
    .add(
        'Props',
        withInfo({ inline: true })(() => {
            return <ButtonGroup />;
        }),
    );
