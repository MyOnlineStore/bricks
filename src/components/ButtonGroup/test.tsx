import React from 'react';
import ButtonGroup from '.';
import Button from '../Button';
import { mountWithTheme } from '../../utility/styled/testing';

describe('ButtonGroup', () => {
    it('renders the correct amount of Buttons', () => {
        const component = mountWithTheme(
            <ButtonGroup>
                <Button title="Test" variant="primary" />
                <Button title="Test" variant="secondary" />
            </ButtonGroup>,
        );

        expect(component.find(Button).length).toBe(2);
    });

    it('renders the correct amount of Buttons when stacked', () => {
        const component = mountWithTheme(
            <ButtonGroup stacked>
                <Button title="Test" variant="primary" />
            </ButtonGroup>,
        );

        expect(component.find(Button).length).toBe(1);
    });

    it('should be testable with a testid', () => {
        const component = mountWithTheme(<ButtonGroup data-testid="buttongroup" />);

        expect(component.find('[data-testid="buttongroup"]').hostNodes()).toHaveLength(1);
    });

    it('should render the buttons in the correct order', () => {
        const component = mountWithTheme(
            <ButtonGroup>
                <Button title="foo" variant="primary" />
                <Button title="bar" variant="secondary" />
            </ButtonGroup>,
        );

        expect(
            component
                .find(Button)
                .last()
                .prop('variant'),
        ).toEqual('primary');

        expect(
            component
                .find(Button)
                .first()
                .prop('variant'),
        ).toEqual('secondary');
    });

    it('should render the buttons in the correct order when stacked', () => {
        const component = mountWithTheme(
            <ButtonGroup stacked>
                <Button title="foo" variant="primary" />
                <Button title="bar" variant="secondary" />
            </ButtonGroup>,
        );

        expect(
            component
                .find(Button)
                .first()
                .prop('variant'),
        ).toEqual('primary');

        expect(
            component
                .find(Button)
                .last()
                .prop('variant'),
        ).toEqual('secondary');
    });
});
