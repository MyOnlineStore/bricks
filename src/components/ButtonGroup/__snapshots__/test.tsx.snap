// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup renders with a Button and a FlatButton in a large node 1`] = `
<styled.div>
  <Box
    key=".0"
  >
    <Button
      title="Test"
      variant="primary"
    />
  </Box>
  <Box
    key=".1"
  >
    <Button
      title="Test"
      variant="secondary"
    />
  </Box>
</styled.div>
`;

exports[`ButtonGroup renders with a single Button in a small node 1`] = `
<styled.div>
  <Box
    key=".0"
  >
    <Button
      title="Test"
      variant="primary"
    />
  </Box>
</styled.div>
`;
